### ISSUE TITLE
2 bugs with paragraphs split between pages

### ISSUE THREAD

**Issue by cdroulers** [pagedjs/pagedjs#167]
I've been working on two really weird bugs where when a paragraph needs to split on the next page, it either disappears completely (even though the markup is there in the elements tab) or is missing two lines before the last

**Comment by cdroulers** [comment]
After messing around a bit more, removing border-bottom on a bunch of elements also seems to fix the issue. Don't know if that helps!

**Comment by julientaq** [comment]
hey @cdroulers sorry i missed those messages (i was lost into a long automated typesetting thingy :D )

so what happens is that pagedjs has some trouble because of how the browser draw things:

padding / margin / border -bottom is added after the element is created.

so in the case of pagedjs, the padding / border / margin appears After the page is cut, so you end up having a bottom offset.

The alternative to that is to use a bit of pagedjs custom class to remove the margins when the element is cut.

in css it looks like this: 

For the bottom of the element that is cut
```
[data-split-to] {padding-bottom: 0}
```

For the top of the element that is cut
```
[data-split-from] {padding-top: 0}
```


tell me if that helps!

**Comment by julientaq** [comment]
(i need to add that to the FAQ when i’m done with the automated thing.)

**Comment by cdroulers** [comment]
Hello @julientaq ! Thank you for the answer!

Using the CSS, but removing border and padding fixes the issues

```css


        [data-split-to] {
            padding-bottom: 0;
            border-bottom: 0;
        }

        [data-split-from] {
            padding-top: 0;
            border-top: 0;
        }
```

This works for now! From what I understand, is that juste a basic CSS to apply to all HTML pages rendered using PagedJS since the issue will usually happen with any element that has padding-bottom or border-bottom?

I'd like to prevent this issue from being triggered by other developers touching our templates, would you have guidance on testing for that kind of issue? Any tips could be welcome!

**Comment by cdroulers** [comment]
There also seems to be some default CSS that tries to remove the padding and margin, but not the borders

<img width="214" alt="image" src="https://github.com/pagedjs/pagedjs/assets/708798/48226d30-3c25-40fb-9459-852e672a5c77">

Version 0.50-beta.0 fixes the issue without custom CSS, but it doesn't seem to be because of the `data-split-*` styles. So it's black magic for me right now. In [this issue](https://gitlab.coko.foundation/pagedjs/pagedjs/-/issues/431#note_126790) you mention that that beta version is not production ready. Is there a timeline for a 0.5.0 release?

**Comment by julientaq** [comment]
oh.

you may want to try this:  https://developer.mozilla.org/en-US/docs/Web/CSS/box-decoration-break

**Comment by cdroulers** [comment]
It seems it worked well when I use `break-inside: avoid;` within my paragraphs, but when I remove it, I can reproduce the issue again.
[super-bug.json.html.txt](https://github.com/pagedjs/pagedjs/files/12935340/super-bug.json.html.txt)
The break-inside was originally to avoid that missing text issue, but I can't find a combination that properly works!

**Comment by cdroulers** [comment]
@julientaq Looks like some parts of text wrap to the top?

<img width="423" alt="image" src="https://github.com/pagedjs/pagedjs/assets/708798/f3d80da9-230a-457e-a18e-24f7685d73d1">

**Comment by julientaq** [comment]
the more i look at it, the more i suspect the `<br>` to put everything down.

wanna try with real `<p>`?

**Comment by julientaq** [comment]
yep.

All talking for that happens in https://mattermost.coko.foundation. 

it’s our main support place.

**Comment by cdroulers** [comment]
Tried with `white-space: pre-line;` and the issue remains, but with `<p>` around each line, it seems to work well for all my issues.

I'll try to rewrite the code that splits user-entered text into lines to use paragraphs and write more tests. Thank you!

**Comment by cdroulers** [comment]
Using `<p>` around each lines seems to work great so far. Am adding lots of tests surrounding the html templates, but will confirm the issue is fixed by removing the `<br>` in the code!

**Comment by NigelCunningham** [comment]
Hi there.

Please see https://github.com/pagedjs/pagedjs/pull/171, which I've just submitted for consideration. It includes a large rework of the pagination code that should help with this issue. I'd appreciate your feedback.

Regards,

Nigel

**Comment by plmunger** [comment]
@julientaq  To add a bit more here, despite what was suggested above, the problem persists for huge paragraphs that span on multiple pages. Here you can see that numbers 1403 to 1450 are in the document but visually missing :

![image](https://github.com/pagedjs/pagedjs/assets/132082264/df85342c-7f5b-4326-9865-1c14bb05c172)

It seemed to me that all that was needed was only to trigger a repaint of the page since I am able to make the missing text appear by toggling some css properties off and then back on in the browser's dev tools (like `box-sizing` for example). Toggling them back on has the effect of reestablishing the page's style to exactly what it was on load, but it doesn't make the text disappear again.

Starting from there I was able to make it work by forcing a repaint of the page with a simple script like this:

```
function forceRepaint() {
  document.body.style.display = "inline";
  setTimeout(() => {
      requestAnimationFrame(() => {
          document.body.style.display = "block";
      });
  }, 0);
}

window.onload = forceRepaint;
window.onbeforeprint =  forceRepaint;
```

**Comment by NigelCunningham** [comment]
I've got a new set of commits I'm about to push on the 171 PR; it's working perfectly for me there now. I'm just checking other linked issues for any other test cases I can try before I push the work.

**Comment by cdroulers** [comment]
@NigelCunningham The original issue contains two reproducible testcases as standalone HTML files, they should be usable pretty easily if you change the <script> to point to your build of pagedjs!

**Comment by NigelCunningham** [comment]
Yeah, I did that and it was working fine with the current set of commits on 171, thanks!

**Comment by EnochGao** [comment]
> Yeah, I did that and it was working fine with the current set of commits on 171, thanks!
My home page header has disappeared
